### ISSUE TITLE
gobetween becomes "stuck" (sometimes)

### ISSUE THREAD

**Issue by shantanugadgil** [yyyar/gobetween#172]
I have GB configured as an SNI router for multiple backend services.

EDIT:
OS: CentOS 7.5+ updates
Docker: Docker version 18.06.1-ce, build e68fc7a
(installed as per instructions on docker.com)

What happens is that after a bit of run time (say a few hours), GB stops responding.
i.e. it accepts the connection, but doesn't forward it to the backend.
I test using `openssl s_client -connect ...`

Restarting the GB docker brings things back to life.

I have been using GB v 0.5.0 (and now 0.6.0)

My LoadBalancer VM is typically a 2 CPU 1 GB machine.

This problem hit again today, so I have made my load balancer VM to 4 CPU 4 GB
(I know for the usage it is way too high).
I am in now wait-and-watch mode if this happens again.

What debug info could I capture to help isolate the problem, the next time this occurs?

FWIW, I found an earlier issue which could be related (not sure though)
https://github.com/yyyar/gobetween/issues/74

Relevant section of the config file:
```
[api]
enabled = true
bind = ":444"

[logging]
level = "debug"
output = "stdout"

[defaults]
protocol = "tcp"
balance = "leastconn"
max_connections = 0
client_idle_timeout = "0"
backend_idle_timeout = "0"
backend_connection_timeout = "0"
##############################################################################

[servers]

[servers.in443]
protocol = "tcp"
bind = ":443"
sni_enabled = true

[servers.in443.sni]
enabled = true
read_timeout = "10s"
hostname_matching_strategy = "regexp"
unexpected_hostname_strategy = "reject"

[servers.in443.discovery]
kind = "consul"
failpolicy = "setempty"
consul_host = "myconsulserver:8500"
# all services should register themselves with this service name
consul_service_name = "in443service"
interval = "60s"
timeout = "10s"

[servers.in443.healthcheck]
kind = "ping"
interval = "60s"
ping_timeout_duration = "5s"
```

Regards,
Shantanu

**Comment by nickdoikov** [comment]
hello @shantanugadgil ,probably you have an issue related to this 
```
client_idle_timeout = "0"
backend_idle_timeout = "0"
```

please provide command output for : 
```
netstat -anp |grep gobetween | grep ESTABLISHED |wc -l > output.txt
netstat -anp |grep gobetween | grep CLOSE_WAIT | wc -l >> output.txt
cat  output.txt
```


and also
```
cat  /proc/$(pidof gobetween)/limits
```

After providing this info - it will be possible to understand does it a Gb bug , or a configuration/sysctl related issue

**Comment by shantanugadgil** [comment]
ok, I'll capture connection counts when this happens again.

Limits:
```
# cat  /proc/$(pidof gobetween)/limits
Limit                     Soft Limit           Hard Limit           Units
Max cpu time              unlimited            unlimited            seconds
Max file size             unlimited            unlimited            bytes
Max data size             unlimited            unlimited            bytes
Max stack size            8388608              unlimited            bytes
Max core file size        unlimited            unlimited            bytes
Max resident set          unlimited            unlimited            bytes
Max processes             unlimited            unlimited            processes
Max open files            1048576              1048576              files
Max locked memory         65536                65536                bytes
Max address space         unlimited            unlimited            bytes
Max file locks            unlimited            unlimited            locks
Max pending signals       15690                15690                signals
Max msgqueue size         819200               819200               bytes
Max nice priority         0                    0
Max realtime priority     0                    0
Max realtime timeout      unlimited            unlimited            us
```

**Comment by shantanugadgil** [comment]
* I added the OS details to the original post.

**Comment by shantanugadgil** [comment]
@nickdoikov  This happened again today.

```
netstat -anp |grep gobetween | grep ESTABLISHED |wc -l > output.txt
netstat -anp |grep gobetween | grep CLOSE_WAIT | wc -l >> output.txt
cat  output.txt
0
0
```

regarding the `idle_timeout` parameters ...

I have one service (among many) which keeps a "forever established" connection with the backend.

What values would you recommend in this case, so that the connection doesn't drop?

**Comment by shantanugadgil** [comment]
Hi,
sorry to report. This happened again after a runtime of about 2 days.

I had modified the config but that hasn't helped:

Modifications I had done.
```
[defaults]
...
backend_connection_timeout = "10s"

[servers.in443.discovery]
consul_service_passing_only = true

[servers.in443.healthcheck]
...
timeout = "60s"
passes = 2
fails = 10

```

When I try an SSL connect from a remote machine, all I see in the logs is the accepted connection.
and the client side gets stuck.

```
2018-09-05 12:54:43 [DEBUG] (server.handle [:443]): Accepted 172.b.c.d:37836 -> [::]:443
```

**Comment by shantanugadgil** [comment]
@illarion
Thanks for this. I use Nomad to launch GB, so I can't run things unattended inside a `screen`.

Thankfully I can switch from using it in `docker` mode to plain `raw_exec` mode. (in Nomad terminology)

I will need to change my loadbalancer job a bit to facilitate this and try again.

**Comment by shantanugadgil** [comment]
hi @illarion 
just a quick question ... is it necessary to run the binary as you described or can it be achieved in the dockerized setup as well?

The "hang" hasn't happened yet, but for testing, I was able to capture a stack trace which looked like
```
gobetween v0.6.0
SIGABRT: abort
PC=0x459643 m=0 sigcode=0

goroutine 21 [syscall]:
runtime.notetsleepg(0x15f13e0, 0x17c248f, 0x0)
        /usr/local/go/src/runtime/lock_futex.go:227 +0x42 fp=0xc420045760 sp=0xc420045730 pc=0x411272
runtime.timerproc(0x15f13c0)
        /usr/local/go/src/runtime/time.go:261 +0x2e7 fp=0xc4200457d8 sp=0xc420045760 pc=0x449477
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc4200457e0 sp=0xc4200457d8 pc=0x4580b1
created by runtime.(*timersBucket).addtimerLocked
        /usr/local/go/src/runtime/time.go:160 +0x107

goroutine 1 [chan receive (nil chan), 1 minutes]:
main.main.func1(0xc420134480)
        /home/ubuntu/git_projects/gobetween/src/main.go:81 +0xde
_/home/ubuntu/git_projects/gobetween/src/cmd.glob..func2(0x159e2e0, 0xc4200336f0, 0x1, 0x1)
        /home/ubuntu/git_projects/gobetween/src/cmd/from-file.go:52 +0x23a
_/home/ubuntu/git_projects/gobetween/src/cmd.glob..func4(0x159e2e0, 0xc4201d6920, 0x0, 0x2)
...
...
...
```

We have planned to put automated capture in place, so that next time the "hang" occurs, the entire stack trace will be captured by sending a kill ABRT on the gobetween machine.

**Comment by shantanugadgil** [comment]
Another one today, different setup, same symptom.
Freeze after about a day's runtime of activity.
[capture_gob_060.log.gz](https://github.com/yyyar/gobetween/files/2369357/capture_gob_060.log.gz)

**Comment by illarion** [comment]
@shantanugadgil how many servers do you have configured?

**Comment by shantanugadgil** [comment]
hi @illarion  by **servers** I assume you means listening ports?

There are about `81` (eighty one) listening ports configured.

`443` is the bulk of the traffic.
Other ports have been setup to forward traffic in ranges to specific backends on specific ports by manipulating the `consul service name`
(e.g. 9001-9020 on gobetween goes to 9001-9020 of a specific backend)

Also, FWIW, I have about 14 (fourteen) distinct backend services behind one GB

**Comment by illarion** [comment]
Ah, ok, got it. I thought that 
```
cat gobetween.capture.1.log | grep -F "tcp.(*Server).Start.func1" | wc -l
80
```
is an indicator of some problem.

**Comment by illarion** [comment]
@shantanugadgil we have a potential bug spot. 

How does this issue affect CPU? We expect it to be constantly ~<number of stuck servers> * 100% (at least one processor should be constantly busy doing busy for loop)

Is that true?

**Comment by illarion** [comment]
@shantanugadgil could you please try gobetween from branch attempt-to-fix-#172 ?

**Comment by illarion** [comment]
@shantanugadgil any updates?

**Comment by shantanugadgil** [comment]
Yes, the build from the branch is being used by a few setup.
Long running tests will reveal if the problem is fixed.
We started on Thursday, I would keep testing for a week.

**Comment by shantanugadgil** [comment]
@illarion 

One more lockup today after about 9 hours.
[gobetween_capture_2018_09_18_12_09_44_log.gz](https://github.com/yyyar/gobetween/files/2393344/gobetween_capture_2018_09_18_12_09_44_log.gz)

**Comment by illarion** [comment]
@shantanugadgil was it gb built from the experimental branch?

**Comment by illarion** [comment]
@shantanugadgil could you please test if server stats work when gb is stuck? And if so, give us the stats please.

**Comment by shantanugadgil** [comment]
@illarion 

the steps:
```
cd work/src/github.com/
cd yyyar/gobetween/
git pull
git checkout attempt-to-fix-#172
<edit VERSION>
make build-static
docker build -t gobetween:f172 .
<upload to local repo and use in test environment>
```

stats work when gb is stuck (I have seen it before), but the current setup is automated to `SIGABRT` the process when passthrough connections stall.

I will try to capture the stats next time this occurs

**Comment by illarion** [comment]
@shantanugadgil it seems that stats are not relevant here. I found the root case (probably).
gb is stuck at reading SNI from the client connection. Investigating now

**Comment by illarion** [comment]
I've added additional check, please rebuild. So:
1. pull the branch
2. rebuild
3. when stuck - do lsof and strace 

Thanks.

**Comment by yyyar** [comment]
@shantanugadgil thank you for your help and commitment to fix this issue.
We're definitely willing to fix it and have progress in understanding the cause, but still need more info as @illarion mentioned above since currently you the only one who gets this issue and can reproduce it. Hope we'll resolve it soon.

**Comment by shantanugadgil** [comment]
Hi @illarion

a nightly run is already under way. I cannot replace the the gb node immediately, though I think I should be able to stop the capture script.

Tomorrow I'll re-build from the same branch `attempt-to-fix-#172` and push out the the new docker build.

Also, I will change the SIGABRT action to `lsof and strace`.

@yyyar 
thanks, I too hope this succeeds! 👍 👍 

P.S.
I hope I don't hit weird Docker and "ptrace" related issues (I will do some trial runs before the next nightly run) 😃

**Comment by shantanugadgil** [comment]
I haven't been able to schedule a run today as I faced build related issues 😢 
(I managed to work around it eventually by doing some steps manually and not via the Makefile, but I guess this will be postponed till tomorrow)

FWIW, during the `make deps` step
```
# go get -v github.com/lxc/lxd/shared
github.com/lxc/lxd/shared
# github.com/lxc/lxd/shared
In file included from vendor/src/github.com/lxc/lxd/shared/network_linux.go:19:0:
./../shared/netns_getifaddrs.c: In function 'get_packet_address':
./../shared/netns_getifaddrs.c:490:2: error: 'for' loop initial declarations are only allowed in C99 mode
  for (unsigned char i = 0; i < n; i++) {
  ^
./../shared/netns_getifaddrs.c:490:2: note: use option -std=c99 or -std=gnu99 to compile your code
``` 

I have the new build now, will push it out tomorrow and see what happens.

**Comment by shantanugadgil** [comment]
A new hang occurred today.

Capture of lsof and strace attached.

[gb.tar.gz](https://github.com/yyyar/gobetween/files/2460236/gb.tar.gz)

**Comment by illarion** [comment]
@shantanugadgil thanks a lot!

It it is a new build from a branch, did you take a look at gobetween log at the time of hang?

**Comment by shantanugadgil** [comment]
After the final fix, the issue has surfaced only once and the long running tests are running fine for days at a time now.

I have not been able to capture the STDOUT while this was going on.

Would it be OK to merge these changes in and roll out a new point version? (and take any further issues from there) ?

Thanks and Regards,
Shantanu

**Comment by illarion** [comment]
Thank you @shantanugadgil , it's merged to master.

**Comment by illarion** [comment]
I will close the issue for now. It is not clear, if it is completely fixed and we could return to it in the future.
